### PR TITLE
Support the MIMO cross-grid path in training loop

### PR DESCRIPTION
**Pull request by yashaswikarnati**
## What
Three adjustments so the stock training loop supports the MIMO cross-grid path, all no-ops for non-MIMO runs:

- `get_forward_backward_func` selects `forward_backward_pipelining_without_interleaving` when passed a `MultiModuleProcessGroupCollection` (the cross-grid bridge); `train()` calls the selector unconditionally instead of branching on `p2p_communicator`.
- Guard the `losses_reduced[0]` access: MIMO encoder-grid ranks are their module's last PP stage but compute no language loss, so the list is empty there.
- Skip the ModelOpt distillation shape hook unless `modelopt_enabled`; it reads `parallel_state`, which is uninitialized on the MIMO path.

## Validation
8-GPU 20L Nemotron VLM e2e: trains + checkpoint save/resume (lm loss 12.18 → 11.54 across resume).

### CODE REVIEW

File: megatron/training/training.py [hunk 1]
```diff
@@ -2388,8 +2391,11 @@ def _save_state_dict(attr_name, label):
     if args.empty_unused_memory_level >= 2:
         torch.cuda.empty_cache()
 
-    if is_last_stage:
-        # Average loss across microbatches.
+    if is_last_stage and losses_reduced:
```
Comment (yashaswikarnati) [inline megatron/training/training.py:2391]: what exactly is last stage is here and how its computed? if its only LLM why we also need losses_reduced ?

File: megatron/training/training.py [hunk 2]
```diff
@@ -2234,8 +2234,11 @@ def train_step(forward_step_func, data_iterator, model, optimizer, opt_param_sch
             model_chunk.force_all_reduce = save_wgrads_in_this_iteration
         optimizer.zero_grad()
 
-        if has_nvidia_modelopt:
-            # [ModelOpt]: Pipeline-parallel Distillation stacks student and teacher tensors
+        if has_nvidia_modelopt and p2p_communicator is None:
```
Comment (yashaswikarnati) [inline megatron/training/training.py]: why did we add  p2p_communicator is None ? when is use_model_opt triggered ?

File: megatron/training/training.py [hunk 3]
```diff
@@ -3352,7 +3358,18 @@ def train(
     eval_duration = 0.0
     eval_iterations = 0
     # Wrap forward_backward_func for Full iteration CUDA graph
-    forward_backward_func = get_forward_backward_func()
+    if p2p_communicator is not None:
+        # MIMO cross-grid bridge: select the schedule by the presence of a
```
Comment (yashaswikarnati) [inline megatron/training/training.py]: verbose doc strings

**Review by Phlip79 — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]

**Review by deepakn94 — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
/ok to test [649065d](https://github.com/NVIDIA/Megatron-LM/pull/5373/commits/649065df094944dbee42708712df2e84f622e4a2)
